### PR TITLE
distribute worker load based on number of blocks loaded

### PR DESCRIPTION
**Pull request by DeveloperInProgress**
# Description
Current method of distribute fetch loads to workers uses the value of heap memory left in a worker as a metric for distribution. This leads to overloading of first worker before fetching begins, or when it is throttled because of a ratelimited endpoint

image

Instead we should use the number of blocks loaded to a worker to distribute further loads. This metric also serves as an approximate comparison for memory used by workers since majority of the heap memory will be occupied by fetched blocks. 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ ] I have tested locally
- [ ] I have performed a self review of my changes
- [ ] Updated any relevant documentation
- [ ] Linked to any relevant issues
- [ ] I have added tests relevant to my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] My code is up to date with the base branch
- [ ] I have updated relevant changelogs. [We suggest using chan](https://github.com/geut/chan/tree/main/packages/chan)

### CODE REVIEW

**Review by stwiname — COMMENTED** [review]
Code looks good but I think this needs tests

File: packages/node-core/src/indexer/blockDispatcher/worker-block-dispatcher.ts [hunk 1]
```diff
@@ -190,9 +192,12 @@ export abstract class WorkerBlockDispatcher<DS, W extends Worker>
   }
 
   private async getNextWorkerIndex(): Promise<number> {
-    return Promise.all(this.workers.map((worker) => worker.getMemoryLeft())).then((memoryLeftValues) => {
-      return memoryLeftValues.indexOf(Math.max(...memoryLeftValues));
-    });
+    this.currentWorkerIndex = (this.currentWorkerIndex + 1) % this.workers.length;
```
Comment (ianhe8x) [inline packages/node-core/src/indexer/blockDispatcher/worker-block-dispatcher.ts:195]: This looks very tricky, you changed a class variable in a `get` method, which only suppose to be readonly.

File: packages/node-core/src/indexer/blockDispatcher/worker-block-dispatcher.ts [hunk 2]
```diff
@@ -190,9 +192,12 @@ export abstract class WorkerBlockDispatcher<DS, W extends Worker>
...
+    const memLeft = await this.workers[this.currentWorkerIndex].getMemoryLeft();
+    if (memLeft < this.minimumHeapLimit) {
+      return this.getNextWorkerIndex();
```
Comment (ianhe8x) [inline packages/node-core/src/indexer/blockDispatcher/worker-block-dispatcher.ts:198]: Can this in certain circumstance become a infinite loop?

**Comment by ianhe8x** [comment]
I think the memory based adjustment on block fetch distribution might end up a less efficiency strategy.

In the beginning, we've tested two strategy, one is distribute jobs to works evenly. But it turned out slower comparing with the other strategy that we always distribute the batch to same worker, plus, when a worker is indexing, it doesn't fetch blocks, so all cpu of a worker can be maximized to spent on indexing rather than splited to block fetching.

Now, this memory based adjustment breaks a batch again.

From my pov, if we are worried about memory usage in block fetching, we can do some control in worker's block fetch logic, pause fetching the blocks in the queue when we find memory is reaching limit.

**Comment by DeveloperInProgress** [comment]
> In the beginning, we've tested two strategy, one is distribute jobs to workers evenly. But it turned out slower comparing with the other strategy that we always distribute a full batch to same worker, plus, when a worker is indexing, it doesn't get more blocks (unless it is the only worker), so all cpu of a worker can be maximized to spent on indexing rather than splited to block fetching.
> 
> Now, this memory based adjustment breaks a batch again.

smart batch sizes would help in the scenarios where a system has limited memory and 

> 
> From my pov, if we are worried about memory usage in block fetching, we can do some control in worker's block fetch logic, pause fetching the blocks in the queue when we find memory is reaching limit.




> @DeveloperInProgress I think the memory based adjustment on block fetch distribution might end up a less efficiency strategy.
> 
> In the beginning, we've tested two strategy, one is distribute jobs to workers evenly. But it turned out slower comparing with the other strategy that we always distribute a full batch to same worker, plus, when a worker is indexing, it doesn't get more blocks (unless it is the only worker), so all cpu of a worker can be maximized to spent on indexing rather than splited to block fetching.
> 
> Now, this memory based adjustment breaks a batch again.
> 
> From my pov, if we are worried about memory usage in block fetching, we can do some control in worker's block fetch logic, pause fetching the blocks in the queue when we find memory is reaching limit.
> 
> `packages/node/src/indexer/blockDispatcher/worker-block-dispatcher.service.ts:L151`

yes it's good to let the GC handle memory management, and we also have a logic to wait for memory if it goes below a limit
https://github.com/subquery/subql/blob/d8383bedca09806bdfb23f94091999b674b74fae/packages/node-core/src/utils/batch-size.ts#L42

On the other hand, should we consider the downsides of loosing smart batch sizing? for example, we would still be on the lookout for OOM error. Sticking to original batch size and waiting for memory to be safe from OOM would increase the idle time.

**Comment by ianhe8x** [comment]
@DeveloperInProgress The safe free memory checks are applied in different layers currently. I think there's a lot of space for simplification and improvement.
Improvements can be in separated PR. For this one, I think my only left concern is the infinite loop it may introduce.

**Review by ianhe8x — APPROVED** [review]

**Pull request merged**